src/speed.c

```diff
@@ -194,72 +122,128 @@ void set_max_end_linear_speed(float value)
 }
 
 /**
- * @brief Set speed parameters with default values from a predefined mode.
+ * @brief Speed turn at a defined speed level.
  *
- * @param[in] mode Speed mode from 0 to (NUM_MODES - 1).
- * @param[in] run Whether or not to use run speed (as opposed to exploration).
+ * @param[in] turn_type Turn type.
+ * @param[in] force Maximum force to apply while turning.
  */
-void set_speed_mode(uint8_t mode, bool run)
+void speed_turn(enum movement turn_type, float force)
 {
-	speed_configuration = mode;
-	linear_acceleration = linear_acceleration_defaults[mode];
-	linear_deceleration = linear_deceleration_defaults[mode];
-	angular_acceleration = angular_acceleration_defaults[mode];
-	if (run)
-		max_linear_speed = max_linear_speed_run_defaults[mode];
-	else
-		max_linear_speed = max_linear_speed_defaults[mode];
-	max_end_linear_speed = max_end_linear_speed_defaults[mode];
+	int32_t start;
+	int32_t current;
+	float travelled;
+	float linear_velocity;
+	float angular_velocity;
+	float max_angular_velocity;
+	float factor;
+	struct turn_parameters turn = turns[turn_type];
+
+	linear_velocity = get_move_turn_linear_speed(turn_type, force);
+	max_angular_velocity = turn.sign * linear_velocity / turn.radius;
+
+	disable_walls_control();
+	start = get_encoder_average_micrometers();
+	while (true) {
+		current = get_encoder_average_micrometers();
+		travelled = (float)(current - start) / MICROMETERS_PER_METER;
+		if (travelled >= 2 * turn.transition + turn.arc)
+			break;
+		angular_velocity = max_angular_velocity;
+		if (travelled < turn.transition) {
+			factor = travelled / turn.transition;
+			angular_velocity *= sin(factor * PI / 2);
+		} else if (travelled >= turn.transition + turn.arc) {
+			factor = (travelled - turn.arc) / turn.transition;
+			angular_velocity *= sin(factor * PI / 2);
+		}
+		set_ideal_angular_speed(angular_velocity);
+	}
+	set_ideal_angular_speed(0);
 }
 
 /**
- * @brief Turn at a defined speed level.
+ * @brief In-place turn at a defined speed level.
  *
- * @param[in] turn Turn type.
- * @param[in] speed Configured speed level.
+ * @param[in] radians Radians to turn (positive means left).
+ * @param[in] force Maximum force to apply while turning.
  */
-void speed_turn(enum movement turn, uint8_t speed)
+void inplace_turn(float radians, float force)
 {
-	parametric_turn(turns[turn][speed].max_angular_speed,
-			turns[turn][speed].rise_time,
-			turns[turn][speed].elapsed_time);
+	int32_t start;
+	int32_t current;
+	float time;
+	float angular_velocity;
+	float max_angular_velocity;
+	float factor;
+	float arc;
+	float transition;
+	float duration;
+	float transition_angle;
+
+	angular_acceleration =
+	    force * MOUSE_WHEELS_SEPARATION / MOUSE_MOMENT_OF_INERTIA;
+	max_angular_velocity = sqrt(radians / 2 * angular_acceleration);
+	if (max_angular_velocity > MOUSE_MAX_ANGULAR_VELOCITY)
+		max_angular_velocity = MOUSE_MAX_ANGULAR_VELOCITY;
+
+	duration = max_angular_velocity / angular_acceleration * PI;
```

Comment:
- Explain formulas on comments
- I don't understand why an inplace turn is so different (we can talk in person :))

Reply:
https://github.com/Bulebots/bulebule/issues/388